compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/TextFieldDelegate.kt

```diff
@@ -126,6 +129,57 @@ internal class TextFieldDelegate {
             TextPainter.paint(canvas, textLayoutResult)
         }
 
+        /**
+         * Notify system that focused input area.
+         *
+         * System is typically scrolled up not to be covered by keyboard.
+         *
+         * @param value The editor model
+         * @param textDelegate The text delegate
+         * @param layoutCoordinates The layout coordinates
+         * @param textInputSession The current input session.
+         * @param hasFocus True if focus is gained.
+         * @param offsetMapping The mapper from/to editing buffer to/from visible text.
+         */
+        @JvmStatic
+        internal fun notifyFocusedRect(
+            value: TextFieldValue,
+            textDelegate: TextDelegate,
+            textLayoutResult: TextLayoutResult,
+            layoutCoordinates: LayoutCoordinates,
+            textInputSession: TextInputSession,
+            hasFocus: Boolean,
+            offsetMapping: OffsetMapping
+        ) {
+            if (!hasFocus) {
+                return
+            }
+            val focusOffsetInTransformed = offsetMapping.originalToTransformed(value.selection.max)
+            val bbox = when {
+                focusOffsetInTransformed < textLayoutResult.layoutInput.text.length -> {
+                    textLayoutResult.getBoundingBox(focusOffsetInTransformed)
+                }
+                focusOffsetInTransformed != 0 -> {
+                    textLayoutResult.getBoundingBox(focusOffsetInTransformed - 1)
+                }
+                else -> { // empty text.
+                    val defaultSize = computeSizeForDefaultText(
+                        textDelegate.style,
+                        textDelegate.density,
+                        textDelegate.fontFamilyResolver
+                    )
+                    Rect(0f, 0f, 1.0f, defaultSize.height.toFloat())
+                }
+            }
+            val globalLT = layoutCoordinates.localToRoot(Offset(bbox.left, bbox.top))
+
+            // TODO remove `Deprecated`, if it is removed in AOSP repository
+            @Suppress("DEPRECATION")
+            textInputSession.notifyFocusedRect(
```

Comment:
Where/how can we observe the focus rect changes? Do we have an API for that?

Reply:
I didn't found such API, but we can write a new one, if we need. Anyway, it seems if we write something like `LocalFocusManager.focusedRect` it will be not what we want. We want the rect for text input. And here we can get it.

Reply:
What I mean is that I see a few places which call `notifyFocusedRect`, but I don't understand which part actually "observers" the changes so input popup position can be set correctly.

Reply:
`notifyFocusedRect` subscribed on 2 things:
- `Modifier.onGloballyPositioned` of the text field, which is called every time its repositioned relative to the window.
- `Modifier.onFocusChanged` which triggers when the focus of text field changes.

I realized, that we don't subscribe on cursor changes. We rely on a side effect of `onGloballyPositioned`, which accidentally is called when we change the cursor position (but it seems, it shouldn't). I will investigate, should we also call `notifyFocusedRect` somewhere else.

Reply:
Call it on relayout too now